From d6fefdc913f0577ddb17dffbe0f6ffea8c3f8d0a Mon Sep 17 00:00:00 2001
From: Kiran Kumar K <kirankumark@marvell.com>
Date: Wed, 1 Mar 2023 17:32:17 +0530
Subject: [PATCH 267/955] common/cnxk: fix spi to sa index destroy

While destroying spi to sa index rule, inline device
mbox should be used. Adding changes to fix this.

Fixes: a3dc6450899 ("common/cnxk: add RTE Flow support for SPI to SA index")

Signed-off-by: Kiran Kumar K <kirankumark@marvell.com>
Change-Id: Ib76b12232129e888af70bb5f45faa55879b89021
Reviewed-on: https://sj1git1.cavium.com/c/IP/SW/dataplane/dpdk/+/98245
Reviewed-by: Nithin Kumar Dabilpuram <ndabilpuram@marvell.com>
Reviewed-by: Jerin Jacob Kollanukkaran <jerinj@marvell.com>
Base-Builds: sa_ip-toolkits-Jenkins <sa_ip-toolkits-jenkins@marvell.com>
Base-Tests: sa_ip-toolkits-Jenkins <sa_ip-toolkits-jenkins@marvell.com>
Tested-by: sa_ip-toolkits-Jenkins <sa_ip-toolkits-jenkins@marvell.com>
---
 drivers/common/cnxk/roc_npc.c                 | 14 ++++++++++----
 .../cnxk-tests/ipsec_msns/cnxk_ipsec_msns.sh  | 19 +++++++++++++++++++
 2 files changed, 29 insertions(+), 4 deletions(-)

diff --git a/drivers/common/cnxk/roc_npc.c b/drivers/common/cnxk/roc_npc.c
index 6bcfd2c6cebd7..e1b56f1b87bf1 100644
--- a/drivers/common/cnxk/roc_npc.c
+++ b/drivers/common/cnxk/roc_npc.c
@@ -1484,16 +1484,22 @@ npc_rss_group_free(struct npc *npc, struct roc_npc_flow *flow)
 static int
 roc_npc_delete_spi_to_sa_action(struct roc_npc *roc_npc, struct roc_npc_flow *flow)
 {
-	struct roc_nix *roc_nix = roc_npc->roc_nix;
 	struct nix_spi_to_sa_delete_req *req;
+	struct nix_inl_dev *inl_dev;
+	struct idev_cfg *idev;
 	struct mbox *mbox;
-	struct nix *nix;
+
+	PLT_SET_USED(roc_npc);
 
 	if (!flow->spi_to_sa_info.has_action || flow->spi_to_sa_info.duplicate)
 		return 0;
 
-	nix = roc_nix_to_nix_priv(roc_nix);
-	mbox = (&nix->dev)->mbox;
+	idev = idev_get_cfg();
+	if (!idev)
+		return -1;
+
+	inl_dev = idev->nix_inl_dev;
+	mbox = inl_dev->dev.mbox;
 	req = mbox_alloc_msg_nix_spi_to_sa_delete(mbox);
 	if (req == NULL)
 		return -ENOSPC;
diff --git a/marvell-ci/test/cnxk-tests/ipsec_msns/cnxk_ipsec_msns.sh b/marvell-ci/test/cnxk-tests/ipsec_msns/cnxk_ipsec_msns.sh
index a26fcfc8c2322..4ee654615d596 100755
--- a/marvell-ci/test/cnxk-tests/ipsec_msns/cnxk_ipsec_msns.sh
+++ b/marvell-ci/test/cnxk-tests/ipsec_msns/cnxk_ipsec_msns.sh
@@ -55,4 +55,23 @@ if [[ $TEST2 != "PASS" ]]; then
 	exit 1
 fi
 
+PART_106B0=$(cat /proc/device-tree/soc\@0/chiprevision)
+
+if [[ $PART_106B0 == "B0" ]]; then
+	TEST3=$(grep "Test RTE_PMD_CNXK_SEC_ACTION_ALG3" $LOG | awk '{print $3}')
+	TEST4=$(grep "Test RTE_PMD_CNXK_SEC_ACTION_ALG4" $LOG | awk '{print $3}')
+
+	if [[ $TEST3 != "PASS" ]]; then
+		echo "Test RTE_PMD_CNXK_SEC_ACTION_ALG3 FAILED"
+		rm -rf $LOG
+		exit 1
+	fi
+
+	if [[ $TEST4 != "PASS" ]]; then
+		echo "Test RTE_PMD_CNXK_SEC_ACTION_ALG4 FAILED"
+		rm -rf $LOG
+		exit 1
+	fi
+fi
+
 echo "TEST SUCCESSFUL"
-- 
2.25.1

